File: GCTP_CPP/Mollweide.cpp
#include "mollweide.h"
#include "projexception.h"

Mollweide::Mollweide(): Pseudocylindrical() 
{
	setName("Mollweide");
	setNumber(MOLL);
}

Mollweide::Mollweide(double gctpParams[], ProjUnit units, ProjDatum dat):
Pseudocylindrical(gctpParams, units, dat) 
{
	setName("Mollweide");
	setNumber(MOLL);
}

void Mollweide::forward_init() 
{
	clearError();

	m_forInitNeeded = false;
}

void Mollweide::inverse_init() 
{
	clearError();

	m_invInitNeeded = false;
}

void Mollweide::forward(double lon, double lat, double* x, double* y) 
{

  double delta_lon;	/* Delta longitude (Given longitude - center */
  double theta;
  double delta_theta;
  double con;
  long i;

  clearError();

  if(m_forInitNeeded)
	  forward_init();

  Util::convertCoords(DEGREE, RADIAN, lon, lat);
  
  /* Forward equations
  -----------------*/
  delta_lon = Util::adjust_lon(lon - m_centerLon);
  theta = lat;
  con = PI * sin(lat);

 /* Iterate using the Newton-Raphson method to find theta
  -----------------------------------------------------*/
  for (i=0;;i++)
	{
	delta_theta = -(theta + sin(theta) - con)/ (1.0 + cos(theta));
	theta += delta_theta;
	if (fabs(delta_theta) < EPSLN) break;
	if (i >= 50) {
		setError(241);
		return;
	}
	  
	}
	theta /= 2.0;

	/*	If the latitude is 90 deg, force the x coordinate to be "0 + false easting"
	this is done here because of precision problems with "cos(theta)"
	 --------------------------------------------------------------------------*/
	if (PI/2 - fabs(lat) < EPSLN)
		delta_lon =0;
	
	m_x_coord = 0.900316316158 * m_radius * delta_lon * cos(theta) + m_falseEasting;
	m_y_coord = 1.4142135623731 * m_radius * sin(theta) + m_falseNorthing;

	Util::convertCoords(METER, m_unitCode, m_x_coord, m_y_coord);

	if(x)
		*x = m_x_coord;
	if(y)
		*y = m_y_coord;
}

void Mollweide::inverse(double x, double y, double* lon, double* lat) 
{
  double theta;
  double arg;

  clearError();

  if(m_invInitNeeded)
	  inverse_init();

  Util::convertCoords(m_unitCode, METER, x, y);

  /* Inverse equations
   -----------------*/
  x -= m_falseEasting; 
  y -= m_falseNorthing;
  arg = y /  (1.4142135623731 * m_radius);

  /* Because of division by zero problems, 'arg' can not be 1.0.  Therefore
     a number very close to one is used instead.
     -------------------------------------------------------------------*/
  if(fabs(arg) > 0.999999999999) 
	  arg=0.999999999999;
  theta = asin(arg);
  m_longitude = Util::adjust_lon(m_centerLon + (x / (0.900316316158 * m_radius * cos(theta))));
  if(m_longitude < (-PI)) 
	  m_longitude = -PI;
  if(m_longitude > PI) 
	  m_longitude = PI;
  arg = (2.0 * theta + sin(2.0 * theta)) / PI;
  if(fabs(arg) > 1.0)arg=1.0;
  m_latitude = asin(arg);

  Util::convertCoords(RADIAN, DEGREE, m_longitude, m_latitude);

  if(lat)
	  *lat = m_latitude;
  if(lon)
	  *lon = m_longitude;

}

File: JGCTP/Mollweide.java
package jgctp;

public class Mollweide extends Projection
{

public 	Mollweide()
{
    super();
    setName("Mollweide");
    setNumber(ProjCode.MOLL);
    
}


public 	Mollweide(double gctpParams[],  ProjUnit units, ProjDatum dat)
{
    super(gctpParams, units, dat);
    setName("Mollweide");
    setNumber(ProjCode.MOLL);
}


protected 	void _init() throws ProjException
{
    return;
}


protected void _forward(GeoPoint p) throws ProjException
{
    double delta_lon;   /* Delta longitude (Given longitude - center */
    double theta;
    double delta_theta;
    double con;
    long i;
    
    /* Forward equations
    -----------------*/
    delta_lon = Util.adjust_lon(p.lon - m_center.lon);
    theta = p.lat;
    con = Constants.PI * Math.sin(p.lat);

   /* Iterate using the Newton-Raphson method to find theta
    -----------------------------------------------------*/
    for (i=0;;i++)
    {
    delta_theta = -(theta + Math.sin(theta) - con)/ (1.0 + Math.cos(theta));
    theta += delta_theta;
    if (Math.abs(delta_theta) < Constants.EPSLN) break;
    if (i >= 50) 
    {
        throw(new ProjException(241, "Mollweide._forward()"));
  
    }
      
    }
    theta /= 2.0;

    /*  If the latitude is 90 deg, force the p.x coordinate to be "0 + false easting"
    this is done here because of precision problems with "Math.cos(theta)"
     --------------------------------------------------------------------------*/
    if (Constants.PI/2 - Math.abs(p.lat) < Constants.EPSLN)
        delta_lon =0;
    
    m_xy.x = 0.900316316158 * m_sphere.radius * delta_lon * Math.cos(theta) + m_falseEasting;
    m_xy.y = 1.4142135623731 * m_sphere.radius * Math.sin(theta) + m_falseNorthing;
}


protected void _inverse(CoordPoint p)
{
    double theta;
    double arg;

    /* Inverse equations
     -----------------*/
    p.x -= m_falseEasting; 
    p.y -= m_falseNorthing;
    arg = p.y /  (1.4142135623731 * m_sphere.radius);

    /* Because of division by zero problems, 'arg' can not be 1.0.  Therefore
       a number very close to one is used instead.
       -------------------------------------------------------------------*/
    if(Math.abs(arg) > 0.999999999999) 
      arg=0.999999999999;
    theta = Math.asin(arg);
    m_lonLat.lon = Util.adjust_lon(m_center.lon + (p.x / (0.900316316158 * m_sphere.radius * Math.cos(theta))));
    if(m_lonLat.lon < (-Constants.PI)) 
      m_lonLat.lon = -Constants.PI;
    if(m_lonLat.lon > Constants.PI) 
      m_lonLat.lon = Constants.PI;
    arg = (2.0 * theta + Math.sin(2.0 * theta)) / Constants.PI;
    if(Math.abs(arg) > 1.0)
      arg=1.0;
    m_lonLat.lat = Math.asin(arg);
}



}